File: questions/Q3362.go
package questions

import (
	"container/heap"
	"sort"
)

func MaxRemoval(nums []int, queries [][]int) int {
	if len(nums) == 0 {
		return -1
	}

	sort.Slice(queries, func(i, j int) bool {
		return queries[i][0] < queries[j][0]
	})
	// Initialize a max heap
	pd := &PriorityQueue{}
	heap.Init(pd)

	currentValue := 0
	differ := make([]int, len(nums)+1)
	queryIndex := 0
	for i := 0; i < len(nums); i++ {
		currentValue += differ[i]
		for queryIndex < len(queries) && queries[queryIndex][0] == i {
			heap.Push(pd, queries[queryIndex][1])
			queryIndex++
		}
		for pd.Len() > 0 && pd.Peek().(int) >= i && currentValue < nums[i] {
			right := heap.Pop(pd)
			differ[right.(int)+1]--
			currentValue++
		}
		if currentValue < nums[i] {
			return -1
		}
	}

	return pd.Len()
}

type PriorityQueue struct {
	values []int
}

// implement heap.Interface
func (pq *PriorityQueue) Push(x interface{}) {
	pq.values = append(pq.values, x.(int))
}

func (pq *PriorityQueue) Pop() interface{} {
	old := pq.values
	n := len(old)
	x := old[n-1]
	pq.values = old[0 : n-1]
	return x
}

func (pq *PriorityQueue) Peek() interface{} {
	if len(pq.values) == 0 {
		return nil
	}
	return pq.values[0]
}

func (pq *PriorityQueue) Less(i, j int) bool {
	return pq.values[i] < pq.values[j]
}

func (pq *PriorityQueue) Swap(i, j int) {
	pq.values[i], pq.values[j] = pq.values[j], pq.values[i]
}

func (pq *PriorityQueue) Len() int {
	return len(pq.values)
}

File: questions/Q3362.java
package questions;

import java.util.Arrays;
import java.util.Collections;
import java.util.PriorityQueue;

public class Q3362 {
    public int maxRemoval(int[] nums, int[][] queries) {
        Arrays.sort(queries, (a, b) -> a[0] - b[0]);
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());

        int[] differ = new int[nums.length+1];
        int value = 0;
        for (int i = 0,j = 0; i < nums.length; i++) {
            value += differ[i];
            for (; j < queries.length; ) {
                if (queries[j][0] == i) {
                    pq.add(queries[j][1]);
                    j++;
                }else {
                    break;
                }
            }
            while (!pq.isEmpty()) {
                int right = pq.poll();
                differ[right+1]--;
                value++;
                if (value == nums[i]) {
                    break;
                }
            }
            if (value < nums[i]) {
                return -1;
            }
        }

        return pq.size();
    }
}
